mllib/src/main/scala/org/apache/spark/mllib/clustering/AffinityPropagation.scala

```diff
@@ -0,0 +1,347 @@
+   * Find the cluster id the given vertex belongs
+   */
+  def findClusterID(vertexID: Long): Option[Int] = {
+    var i = 0
+    clusters.foreach(cluster => {
+      if (cluster.contains(vertexID)) {
+        return Some(i)
+      }
+      i += i
+    })
+    None 
+  } 
+}
+
+/**
+ * :: Experimental ::
+ *
+ * AffinityPropagation (AP), a graph clustering algorithm based on the concept of "message passing"
+ * between data points. Unlike clustering algorithms such as k-means or k-medoids, AP does not
+ * require the number of clusters to be determined or estimated before running it. AP is developed
+ * by [[http://www.psi.toronto.edu/affinitypropagation/FreyDueckScience07.pdf Frey and Dueck]].
+ *
+ * @param maxIterations Maximum number of iterations of the AP algorithm.
+ *
+ * @see [[http://en.wikipedia.org/wiki/Affinity_propagation (Wikipedia)]]
+ */
+@Experimental
+class AffinityPropagation private[clustering] (
+    private var maxIterations: Int,
+    private var lambda: Double,
+    private var normalization: Boolean) extends Serializable {
+
+  import org.apache.spark.mllib.clustering.AffinityPropagation._
+
+  /** Constructs a AP instance with default parameters: {maxIterations: 100, lambda: 0.5,
+   *    normalization: false}.
+   */
+  def this() = this(maxIterations = 100, lambda = 0.5, normalization = false)
+
+  /**
+   * Set maximum number of iterations of the messaging iteration loop
+   */
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+ 
+  /**
+   * Get maximum number of iterations of the messaging iteration loop
+   */
+  def getMaxIterations(): Int = {
+    this.maxIterations
+  }
+ 
+  /**
+   * Set lambda of the messaging iteration loop
+   */
+  def setLambda(lambda: Double): this.type = {
+    this.lambda = lambda
+    this
+  }
+ 
+  /**
+   * Get lambda of the messaging iteration loop
+   */
+  def getLambda(): Double = {
+    this.lambda
+  }
+ 
+  /**
+   * Set whether to do normalization or not
+   */
+  def setNormalization(normalization: Boolean): this.type = {
+    this.normalization = normalization
+    this
+  }
+ 
+  /**
+   * Get whether to do normalization or not
+   */
+  def getNormalization(): Boolean = {
+    this.normalization
+  }
+ 
+  /**
+   * Run the AP algorithm.
+   *
+   * @param similarities an RDD of (i, j, s,,ij,,) tuples representing the similarity matrix, which
+   *                     is the matrix S in the AP paper. The similarity s,,ij,, is set to
+   *                     real-valued similarities. This is not required to be a symmetric matrix 
+   *                     and hence s,,ij,, can be not equal to s,,ji,,. Tuples with i = j are
+   *                     referred to as "preferences" in the AP paper. The data points with larger
+   *                     values of s,,ii,, are more likely to be chosen as exemplars.
+   *
+   * @param symmetric the given similarity matrix is symmetric or not. Default value: true
```

Comment:
Do you mean to set `symmetric ` as a parameter of `AffinityPropagation`? Because it is closely related to the characters of the given `similarities`, I list it as a parameter of `run` method. But I don't feel strong about it.

I think if it is better to use other name instead of `symmetric`. This parameter indicates that `(i, j)` equals to `(j, i)`, so in order to avoid redundancy, the given `similarities` only cover triangular matrix. May just use `triangular`?